---
title: "Grey Jay Core Microbiome"
author: "Alicia Halhed"
date: "30/05/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Set Up
```{r}
# set working directory
setwd("~/OneDrive - University of Guelph/Alicia's Thesis/GreyJayMicrobiome/")
# attach required packages
library(qiime2R)
library(vegan)
library(tidyverse)
# set theme
theme_set(theme_bw())
```

```{r, message=FALSE}
# data
nReads <- 344
otu <- read_qza("~/OneDrive - University of Guelph/Alicia's Thesis/GreyJayMicrobiome/rarefied-table.qza")$data
map <- read_tsv("~/OneDrive - University of Guelph/Alicia's Thesis/GreyJayMicrobiome/input/jay-met.tsv")

# transforms
otu_PA <- 1*((otu>0)==1)                                               # presence-absence data
otu_occ <- rowSums(otu_PA)/ncol(otu_PA)                                # occupancy calculation
otu_rel <- apply(decostand(otu, method="total", MARGIN=2),1, mean)     # mean relative abundance
occ_abun <- rownames_to_column(as.data.frame(cbind(otu_occ, otu_rel)),'otu') # combining occupancy and abundance data frame
```

```{r}
# Ranking OTUs based on their occupancy
# For caluclating raking index we included following conditions:
#   - time-specific occupancy (sumF) = frequency of detection within time point (genotype or site)
#   - replication consistency (sumG) = has occupancy of 1 in at least one time point (genotype or site) (1 if occupancy 1, else 0)

PresenceSum <- data.frame(otu = as.factor(row.names(otu)), otu) %>% 
  gather(sampleid, abun, -otu) %>%
  left_join(map, by = 'sampleid') %>%
  group_by(otu, CollectionDate) %>%
  summarise(time_freq=sum(abun>0)/length(abun),            # frequency of detection between time points
            coreTime=ifelse(time_freq == 1, 1, 0)) %>%     # 1 only if occupancy 1 with specific time, 0 if not
  group_by(otu) %>%
  summarise(sumF=sum(time_freq),
            sumG=sum(coreTime),
            nS=length(CollectionDate)*2,           
            Index=(sumF+sumG)/nS)                 # calculating weighting Index based on number of time points detected and 

otu_ranked <- occ_abun %>%
  left_join(PresenceSum, by='otu') %>%
  transmute(otu=otu,
            rank=Index) %>%
  arrange(desc(rank))
```


```{r}
# Calculating the contribution of ranked OTUs to the BC similarity
BCaddition <- NULL

# calculating BC dissimilarity based on the 1st ranked OTU
otu_start=otu_ranked$otu[1]                   
start_matrix <- as.matrix(otu[otu_start,])
start_matrix <- t(start_matrix)
x <- apply(combn(ncol(start_matrix), 2), 2, function(x) sum(abs(start_matrix[,x[1]]- start_matrix[,x[2]]))/(2*nReads))
x_names <- apply(combn(ncol(start_matrix), 2), 2, function(x) paste(colnames(start_matrix)[x], collapse=' - '))
df_s <- data.frame(x_names,x)
names(df_s)[2] <- 1 
BCaddition <- rbind(BCaddition,df_s)
# calculating BC dissimilarity based on additon of ranked OTUs from 2nd to 500th. Can be set to the entire length of OTUs in the dataset, however it might take some time if more than 5000 OTUs are included.
for(i in 2:647){                              
  otu_add=otu_ranked$otu[i]                       
  add_matrix <- as.matrix(otu[otu_add,])
  add_matrix <- t(add_matrix)
  start_matrix <- rbind(start_matrix, add_matrix)
  x <- apply(combn(ncol(start_matrix), 2), 2, function(x) sum(abs(start_matrix[,x[1]]-start_matrix[,x[2]]))/(2*nReads))
  x_names <- apply(combn(ncol(start_matrix), 2), 2, function(x) paste(colnames(start_matrix)[x], collapse=' - '))
  df_a <- data.frame(x_names,x)
  names(df_a)[2] <- i 
  BCaddition <- left_join(BCaddition, df_a, by=c('x_names'))
}
# calculating the BC dissimilarity of the whole dataset (not needed if the second loop is already including all OTUs) 
x <-  apply(combn(ncol(otu), 2), 2, function(x) sum(abs(otu[,x[1]]-otu[,x[2]]))/(2*nReads))   
x_names <- apply(combn(ncol(otu), 2), 2, function(x) paste(colnames(otu)[x], collapse=' - '))
df_full <- data.frame(x_names,x)
names(df_full)[2] <- length(rownames(otu))
BCfull <- left_join(BCaddition,df_full, by='x_names')

rownames(BCfull) <- BCfull$x_names
temp_BC <- BCfull
temp_BC$x_names <- NULL
temp_BC_matrix <- as.matrix(temp_BC)

BC_ranked <- data.frame(rank = as.factor(row.names(t(temp_BC_matrix))),t(temp_BC_matrix)) %>% 
  gather(comparison, BC, -rank) %>%
  group_by(rank) %>%
  summarise(MeanBC=mean(BC)) %>%            # mean Bray-Curtis dissimilarity
  arrange(desc(-MeanBC)) %>%
  mutate(proportionBC=MeanBC/max(MeanBC))   # proportion of the dissimilarity explained by the n number of ranked OTUs
Increase=BC_ranked$MeanBC[-1]/BC_ranked$MeanBC[-length(BC_ranked$MeanBC)]
increaseDF <- data.frame(IncreaseBC=c(0,(Increase)), rank=factor(c(1:(length(Increase)+1))))
BC_ranked <- left_join(BC_ranked, increaseDF)
BC_ranked <- BC_ranked[-nrow(BC_ranked),]
```

```{r Plot}
#Creating thresholds for core inclusion 

#Method: 
#A) Elbow method (first order difference) (script modified from https://pommevilla.github.io/random/elbows.html)
fo_difference <- function(pos){
  left <- (BC_ranked[pos, 2] - BC_ranked[1, 2]) / pos
  right <- (BC_ranked[nrow(BC_ranked), 2] - BC_ranked[pos, 2]) / (nrow(BC_ranked) - pos)
  return(left - right)
}
BC_ranked$fo_diffs <- sapply(1:nrow(BC_ranked), fo_difference)

elbow <- which.max(BC_ranked$fo_diffs)

#B) Final increase in BC similarity of equal or greater then 2% 
lastCall <- last(as.numeric(BC_ranked$rank[(BC_ranked$IncreaseBC>=1.02)]))

#Creating occupancy abundance plot
occ_abun$fill <- 'no'
# correcting the NA issue with the cbinds
occ_abun$fill[occ_abun$otu %in% cbind(BC_ranked, otu_ranked)$otu[cbind(BC_ranked, otu_ranked)$IncreaseBC>=1.02]] <- 'core'
```

```{r}
# add taxonomy information to occ_abun
tax <- read_qza("~/OneDrive - University of Guelph/Alicia's Thesis/GreyJayMicrobiome/taxonomy/silva-taxonomy.qza")$data
# clean up/separatee taxonomy labels
tax$Taxon <- tax$Taxon %>%
  str_replace_all("d__", "") %>%
  str_replace_all("p__", "") %>%
  str_replace_all("c__", "") %>%
  str_replace_all("o__", "") %>%
  str_replace_all("f__", "") %>%
  str_replace_all("g__", "") %>%
  str_replace_all("s__", "")
occ_abunT <- tax %>% 
  mutate("Taxon" = gsub(".*__", "", .$Taxon),
         "Kingdom" = word(.$Taxon, 1, sep = ";"),
         "Phylum" = word(.$Taxon, 2, sep = ";"),
         "Class" = word(.$Taxon, 3, sep = ";"),
         "Order" = word(.$Taxon, 4, sep = ";"),
         "Family" = word(.$Taxon, 5, sep = ";"),
         "Genus" = word(.$Taxon, 6, sep = ";"),
         "Species" = word(.$Taxon, 7, sep = ";")) %>%
  # join to core labels
  right_join(occ_abun, by = c("Feature.ID" = "otu" ))
# save core information to file
write.table(occ_abunT, file = "~/OneDrive - University of Guelph/Alicia's Thesis/GreyJayMicrobiome/CanadaJayMicrobiome/data/coreJay.csv", sep = ",", quote = F, row.names = F)
#View(occ_abunT)
```


```{r}
# plot
ggplot(occ_abun, aes(y = otu_occ, x = otu_rel, color = fill)) + 
  geom_point() +
  # log transform the x axis
  scale_x_log10() +
  # set discrete viridis colour scheme
  scale_colour_viridis_d() + 
  # add axis labels
  labs(x = "Mean Relative Abundance of Each OTU (log10)", y = "Occupancy (Proportion of Samples)")
```

